### PR TITLE
meson: 0.60.3 -> 0.61.2

### PR DESCRIPTION
**Pull request by NickCao**
###### Motivation for this change
ref: https://github.com/NixOS/nixpkgs/pull/154708

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by bobby285271** [comment]
Bumping meson to 0.61.0 will break most Pantheon packages without proper patching, then block the nixos-unstable channel (mostly fails with "Function does not take positional arguments". Yeah this is fixed on upstream already, but I don't think elementary will make some release before elementary OS 7), if we have to do this, I think I should open a PR to fetchpatch those fixes first.

**Comment by jtojnar** [comment]
Yeah, it is in the original PR: https://github.com/NixOS/nixpkgs/pull/144779/files#diff-f4a18ab9c966d0274b79b396e0f506406ee5fb7d60758f71bc3de9662db8d455

**Comment by NickCao** [comment]
gobject-introspection is also broken: https://gitlab.gnome.org/GNOME/gobject-introspection/-/commit/effb1e09dee263cdac4ec593e8caf316e6f01fe2

**Comment by NickCao** [comment]
harfbuzz breaks when cross building docs:
```
harfbuzz-riscv64-unknown-linux-gnu>     t_args.append('--run=' + ' '.join(state.environment.get_exe_wrapper().get_command()))
harfbuzz-riscv64-unknown-linux-gnu> AttributeError: 'NoneType' object has no attribute 'get_command'
harfbuzz-riscv64-unknown-linux-gnu> docs/meson.build:50:6: ERROR: Unhandled python exception
harfbuzz-riscv64-unknown-linux-gnu>     This is a Meson bug and should be reported!
```
can be worked around by disabling docs when crossing.

**Comment by NickCao** [comment]
applied multiple patches following archlinux, ref: https://github.com/archlinux/svntogit-packages/blob/31d9974f1c4d3badca5809d74bc8c5e8379e3ee3/trunk/PKGBUILD

**Comment by Mindavi** [comment]
Yeah, either this or at least adding the cross-compilation patch would be nice. Adding the patch is probably 'safer' for now, but I don't know what the breakage for the meson bump actually is. Reading this thread it might be painful regardless of what we choose.

**Comment by alyssais** [comment]
Yeah, I think we should just apply the patch like the original PR did to unblock staging next, and then retarget this to staging where there's a bit less time pressure.

**Comment by NickCao** [comment]
meson 0.60 is actually as broken as 0.61, and there are already enough upstream fixes to apply at least for the critical packages. I'm more in favour for skip 0.60.

**Comment by jtojnar** [comment]
Not sure what you mean. Both glib and gobject-introspection seem to build in native compilation with 0.60.

**Comment by NickCao** [comment]
> Not sure what you mean. Both glib and gobject-introspection seem to build in native compilation with 0.60.

My mistake, you are correct, let's just do as @alyssais said, bring back the lost patch and retarget this one.

**Comment by jonringer** [comment]
is this in a reviewable state?

File: pkgs/development/tools/build-managers/meson/default.nix
Comment (jtojnar) [inline pkgs/development/tools/build-managers/meson/default.nix:63]: Remove the file as well?

**Comment by NickCao** [comment]
> is this in a reviewable state?

yes, still have to tackle with a bunch of build failures in the next staging-next cycle though, but meson itself should be good as is.

**Comment by jtojnar** [comment]
polkit https://gitlab.freedesktop.org/polkit/polkit/-/merge_requests/105

**Comment by alyssais** [comment]
@jtojnar @AndersonTorres are you happy with this? The next meson release is scheduled for tomorrow, so would be great if we could get this in so we don't have to upgrade through multiple versions at once again.

**Comment by NickCao** [comment]
Why not just wait for another day? Skipping a meson release looks like a benefit for us.

**Comment by AndersonTorres** [comment]
If the next Meson is not a 0.62.x, then wait until tomorrow.

**Comment by NickCao** [comment]
The newly tagged meson 0.62.0rc1 breaks glib even in native compilation, let's get this in staging and wait for the new meson release to brew.

**Comment by AndersonTorres** [comment]
RC release? 
Breaking even the easiest case?

Oh man...

**Review by AndersonTorres — APPROVED** [review]

**Comment by jtojnar** [comment]
polkit is still broken:

```
actions/meson.build:3:5: ERROR: Function does not take positional arguments.
```

**Comment by AndersonTorres** [comment]
Why is Darwin failing?

**Comment by NickCao** [comment]
> Why is Darwin failing?

Looks like dependency failures?

**Comment by AndersonTorres** [comment]
That being said, I will merge this.

**Comment by trofi** [comment]
`meld` and `gcr` are also affected:

```
error: builder for '/nix/store/c10z65p2xbbawp2rkf4ckldkxw2vhb27-meld-3.21.1.drv' failed with exit code 1;
       last 10 log lines:
       > Configuring conf.py using configuration
       > Found pkg-config: /nix/store/hmqp1izgl9w66hi498g2yra9blxgkahi-pkg-config-wrapper-0.29.2/bin/pkg-config (0.29.2)
       > Program glib-compile-resources found: YES (/nix/store/s9y11yw1m3ymi29phfzgj589dgnqyn1r-glib-2.70.3-dev/bin/glib-compile-resources)
       > Configuring org.gnome.Meld.desktop.in using configuration
       > Program desktop-file-validate found: YES (/nix/store/ihlxnklqbi2c9862z1vq3rr7ay9i5gxx-desktop-file-utils-0.26/bin/desktop-file-validate)
       > Configuring org.gnome.meld.appdata.xml.in using configuration
       >
       > data/meson.build:34:0: ERROR: Function does not take positional arguments.
       >
       > A full log can be found at /build/meld-3.21.1/build/meson-logs/meson-log.txt
       For full logs, run 'nix log /nix/store/c10z65p2xbbawp2rkf4ckldkxw2vhb27-meld-3.21.1.drv'.
```

```
error: builder for '/nix/store/3zfnzk2xlq1nj62h80rks6m8vc7g4x40-gcr-3.41.0.drv' failed with exit code 1;
       last 10 log lines:
       > Configuring gck-version.h using configuration
       > Program glib-mkenums found: YES (/nix/store/s9y11yw1m3ymi29phfzgj589dgnqyn1r-glib-2.70.3-dev/bin/glib-mkenums)
       > Program glib-mkenums found: YES (/nix/store/s9y11yw1m3ymi29phfzgj589dgnqyn1r-glib-2.70.3-dev/bin/glib-mkenums)
       > Found pkg-config: /nix/store/hmqp1izgl9w66hi498g2yra9blxgkahi-pkg-config-wrapper-0.29.2/bin/pkg-config (0.29.2)
       > Program glib-genmarshal found: YES (/nix/store/s9y11yw1m3ymi29phfzgj589dgnqyn1r-glib-2.70.3-dev/bin/glib-genmarshal)
       > Compiler for C supports link arguments -Wl,--version-script,/build/gcr-3.41.0/gck/libgck.map: YES
       >
       > gck/meson.build:130:2: ERROR: gnome.generate_gir got unknown keyword arguments "packages"
       >
       > A full log can be found at /build/gcr-3.41.0/build/meson-logs/meson-log.txt
       For full logs, run 'nix log /nix/store/3zfnzk2xlq1nj62h80rks6m8vc7g4x40-gcr-3.41.0.drv'.
```

**Comment by bobby285271** [comment]
Maybe we will need [this patch](https://gitlab.gnome.org/GNOME/gcr/-/commit/b3ca1d02bb0148ca787ac4aead164d7c8ce2c4d8) for gcr.

For `Function does not take positional arguments.` we are thinking of patching meson with https://github.com/mesonbuild/meson/commit/022c0357ae4cd6c1d7e9a0f3a2672ba410a71fa2, some discussion happens in #164266.

**Comment by NickCao** [comment]
When would the next staging-next cycle begin? I would like to take a look at the blast radius before determining whether it's appropriate to patch meson (again). FYI, meson 0.61.3 was just released 2 days ago, nothing interesting though, jut a reminder.

**Comment by vcunat** [comment]
It should be as soon as Hydra recovers from the openssl update.

**Comment by trofi** [comment]
Proposed `gcr` fix as https://github.com/NixOS/nixpkgs/pull/164945

**Comment by trofi** [comment]
Proposed `meld` fix as https://github.com/NixOS/nixpkgs/pull/164948

**Comment by Mindavi** [comment]
Harfbuzz cross-compilation seems to have broken in this update, might investigate later.

```
Program gtkdoc-fixxref found: YES (/nix/store/yglq5bwhn1j3qpgv06hv4nbvgjinwik4-gtk-doc-1.33.2/bin/gtkdoc-fixxref)
Traceback (most recent call last):
  File "/nix/store/54pqyk0ni1c1z6ywiba4vf30pa1q8rr8-meson-0.61.2/lib/python3.9/site-packages/mesonbuild/mesonmain.py", line 146, in run
    return options.run_func(options)
  File "/nix/store/54pqyk0ni1c1z6ywiba4vf30pa1q8rr8-meson-0.61.2/lib/python3.9/site-packages/mesonbuild/msetup.py", line 294, in run
    app.generate()
  File "/nix/store/54pqyk0ni1c1z6ywiba4vf30pa1q8rr8-meson-0.61.2/lib/python3.9/site-packages/mesonbuild/msetup.py", line 185, in generate
    self._generate(env)
  File "/nix/store/54pqyk0ni1c1z6ywiba4vf30pa1q8rr8-meson-0.61.2/lib/python3.9/site-packages/mesonbuild/msetup.py", line 229, in _generate
    intr.run()
  File "/nix/store/54pqyk0ni1c1z6ywiba4vf30pa1q8rr8-meson-0.61.2/lib/python3.9/site-packages/mesonbuild/interpreter/interpreter.py", line 2549, in run
    super().run()
  File "/nix/store/54pqyk0ni1c1z6ywiba4vf30pa1q8rr8-meson-0.61.2/lib/python3.9/site-packages/mesonbuild/interpreterbase/interpreterbase.py", line 148, in run
    self.evaluate_codeblock(self.ast, start=1)
  File "/nix/store/54pqyk0ni1c1z6ywiba4vf30pa1q8rr8-meson-0.61.2/lib/python3.9/site-packages/mesonbuild/interpreterbase/interpreterbase.py", line 173, in evaluate_codeblock
    raise e
  File "/nix/store/54pqyk0ni1c1z6ywiba4vf30pa1q8rr8-meson-0.61.2/lib/python3.9/site-packages/mesonbuild/interpreterbase/interpreterbase.py", line 166, in evaluate_codeblock
    self.evaluate_statement(cur)
  File "/nix/store/54pqyk0ni1c1z6ywiba4vf30pa1q8rr8-meson-0.61.2/lib/python3.9/site-packages/mesonbuild/interpreterbase/interpreterbase.py", line 189, in evaluate_statement
    return self.evaluate_if(cur)
  File "/nix/store/54pqyk0ni1c1z6ywiba4vf30pa1q8rr8-meson-0.61.2/lib/python3.9/site-packages/mesonbuild/interpreterbase/interpreterbase.py", line 272, in evaluate_if
    self.evaluate_codeblock(i.block)
  File "/nix/store/54pqyk0ni1c1z6ywiba4vf30pa1q8rr8-meson-0.61.2/lib/python3.9/site-packages/mesonbuild/interpreterbase/interpreterbase.py", line 173, in evaluate_codeblock
    raise e
  File "/nix/store/54pqyk0ni1c1z6ywiba4vf30pa1q8rr8-meson-0.61.2/lib/python3.9/site-packages/mesonbuild/interpreterbase/interpreterbase.py", line 166, in evaluate_codeblock
    self.evaluate_statement(cur)
  File "/nix/store/54pqyk0ni1c1z6ywiba4vf30pa1q8rr8-meson-0.61.2/lib/python3.9/site-packages/mesonbuild/interpreterbase/interpreterbase.py", line 179, in evaluate_statement
    return self.function_call(cur)
  File "/nix/store/54pqyk0ni1c1z6ywiba4vf30pa1q8rr8-meson-0.61.2/lib/python3.9/site-packages/mesonbuild/interpreterbase/interpreterbase.py", line 454, in function_call
    res = func(node, func_args, kwargs)
  File "/nix/store/54pqyk0ni1c1z6ywiba4vf30pa1q8rr8-meson-0.61.2/lib/python3.9/site-packages/mesonbuild/interpreterbase/decorators.py", line 276, in wrapper
    return f(*nargs, **wrapped_kwargs)
  File "/nix/store/54pqyk0ni1c1z6ywiba4vf30pa1q8rr8-meson-0.61.2/lib/python3.9/site-packages/mesonbuild/interpreterbase/decorators.py", line 573, in wrapper
    return f(*wrapped_args, **wrapped_kwargs)
  File "/nix/store/54pqyk0ni1c1z6ywiba4vf30pa1q8rr8-meson-0.61.2/lib/python3.9/site-packages/mesonbuild/interpreter/interpreter.py", line 2039, in func_subdir
    self.evaluate_codeblock(codeblock)
  File "/nix/store/54pqyk0ni1c1z6ywiba4vf30pa1q8rr8-meson-0.61.2/lib/python3.9/site-packages/mesonbuild/interpreterbase/interpreterbase.py", line 173, in evaluate_codeblock
    raise e
  File "/nix/store/54pqyk0ni1c1z6ywiba4vf30pa1q8rr8-meson-0.61.2/lib/python3.9/site-packages/mesonbuild/interpreterbase/interpreterbase.py", line 166, in evaluate_codeblock
    self.evaluate_statement(cur)
  File "/nix/store/54pqyk0ni1c1z6ywiba4vf30pa1q8rr8-meson-0.61.2/lib/python3.9/site-packages/mesonbuild/interpreterbase/interpreterbase.py", line 183, in evaluate_statement
    return self.method_call(cur)
  File "/nix/store/54pqyk0ni1c1z6ywiba4vf30pa1q8rr8-meson-0.61.2/lib/python3.9/site-packages/mesonbuild/interpreterbase/interpreterbase.py", line 482, in method_call
    res = obj.method_call(method_name, args, kwargs)
  File "/nix/store/54pqyk0ni1c1z6ywiba4vf30pa1q8rr8-meson-0.61.2/lib/python3.9/site-packages/mesonbuild/interpreter/interpreterobjects.py", line 761, in method_call
    ret = method(state, args, kwargs)
  File "/nix/store/54pqyk0ni1c1z6ywiba4vf30pa1q8rr8-meson-0.61.2/lib/python3.9/site-packages/mesonbuild/interpreterbase/decorators.py", line 276, in wrapper
    return f(*nargs, **wrapped_kwargs)
  File "/nix/store/54pqyk0ni1c1z6ywiba4vf30pa1q8rr8-meson-0.61.2/lib/python3.9/site-packages/mesonbuild/interpreterbase/decorators.py", line 573, in wrapper
    return f(*wrapped_args, **wrapped_kwargs)
  File "/nix/store/54pqyk0ni1c1z6ywiba4vf30pa1q8rr8-meson-0.61.2/lib/python3.9/site-packages/mesonbuild/modules/gnome.py", line 1365, in gtkdoc
    t_args.append('--run=' + ' '.join(state.environment.get_exe_wrapper().get_command()))
AttributeError: 'NoneType' object has no attribute 'get_command'

docs/meson.build:50:6: ERROR: Unhandled python exception

    This is a Meson bug and should be reported!
```

Probably an upstream bug.

**Pull request merged**
